### PR TITLE
PMM-4906 PMM-4541 skip system dbs and collections, decrease log level for some messages, refactoring

### PR DESCRIPTION
**Pull request by artemgavrilov**
1. Skip system databases and collections during the scan
2. Decease log level of non-critical messages
3. Refactoring

Based on https://github.com/percona/mongodb_exporter/pull/122

### CODE REVIEW

**Comment by akira-kurogane** [comment]
No, I don't like this.
* The MongoDB System collections are not for monitoring. Nobody needs to monitor them. An optional flag is not the right way to prevent new users from getting in it and raising new tickets; or just rejecting PMM with MongoDB outright.
* Changing the log level doesn't reduce log bloat. It also still causes all the lines in the mongod log file. Which can lead to disk-full incidents that take the replicaset down.

Please apply PR #122 instead.

I haven't looked at the refactoring in the patch here.

**Comment by akira-kurogane** [comment]
Thanks Aleksi, Artem.

File: .golangci.yml [hunk 1]
```diff
@@ -0,0 +1,41 @@
+linters-settings:
+  depguard:
+    list-type: blacklist
+    include-go-root: true
+    packages:
+      # use "github.com/pkg/errors" instead
+      - errors
+      # use "github.com/golang/protobuf/proto" instead
+      - github.com/gogo/protobuf/proto
```
Comment (AlekSi) [inline .golangci.yml]: mongodb_exporter does not use protobuf.

File: collector/common/collection.go [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+// CollFullName returns collections full name
```
Comment (AlekSi) [inline collector/common/collection.go]: Please add dots to full sentence comments:
```
// CollFullName returns collection full name.
```
but
```
// fast-path
```

File: collector/common/system.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+	configDB = "config"
+	localDB  = "local"
+)
+
+const sysCollPrefix = "system."
+
+// IsSystemDB tests whether system db name passed
+func IsSystemDB(dbName string) bool {
+	switch dbName {
+	case adminDB, configDB, localDB:
```
Comment (AlekSi) [inline collector/common/system.go]: If those strings are used only in once place, there is no point in replacing string literals with constants with the same names.

File: collector/mongod/collections_status.go [hunk 4]
```diff
@@ -121,64 +122,58 @@ func (collStatList *CollectionStatList) Describe(ch chan<- *prometheus.Desc) {
 	collectionIndexesSize.Describe(ch)
 }
 
-var (
-	logSuppressCS = make(map[string]bool)
-)
+var logSuppressCS = make(map[string]struct{})
```
Comment (AlekSi) [inline collector/mongod/collections_status.go]: Access to that global variable is not synchronized:

```
==================
WARNING: DATA RACE
Write at 0x00c00015f7a0 by goroutine 120:
  runtime.mapdelete_faststr()
      /usr/local/Cellar/go/1.13.8/libexec/src/runtime/map_faststr.go:297 +0x0
  github.com/percona/mongodb_exporter/collector/mongod.GetCollectionStatList()
      /Users/aleksi/Code/Percona/PMM-2.0/src/github.com/percona/mongodb_exporter/collector/mongod/collections_status.go:141 +0x160
  github.com/percona/mongodb_exporter/collector.(*MongodbCollector).collectMongod()
      /Users/aleksi/Code/Percona/PMM-2.0/src/github.com/percona/mongodb_exporter/collector/mongodb_collector.go:277 +0x36e
  github.com/percona/mongodb_exporter/collector.(*MongodbCollector).scrape()
      /Users/aleksi/Code/Percona/PMM-2.0/src/github.com/percona/mongodb_exporter/collector/mongodb_collector.go:213 +0x69a
  github.com/percona/mongodb_exporter/collector.(*MongodbCollector).Collect()
      /Users/aleksi/Code/Percona/PMM-2.0/src/github.com/percona/mongodb_exporter/collector/mongodb_collector.go:162 +0x46
  github.com/prometheus/client_golang/prometheus.(*Registry).Gather.func1()
      /Users/aleksi/Soft/GOPATH/pkg/mod/github.com/prometheus/client_golang@v1.0.0/prometheus/registry.go:442 +0x1f5

Previous write at 0x00c00015f7a0 by goroutine 121:
  runtime.mapdelete_faststr()
      /usr/local/Cellar/go/1.13.8/libexec/src/runtime/map_faststr.go:297 +0x0
  github.com/percona/mongodb_exporter/collector/mongod.GetCollectionStatList()
      /Users/aleksi/Code/Percona/PMM-2.0/src/github.com/percona/mongodb_exporter/collector/mongod/collections_status.go:141 +0x160
  github.com/percona/mongodb_exporter/collector.(*MongodbCollector).collectMongod()
      /Users/aleksi/Code/Percona/PMM-2.0/src/github.com/percona/mongodb_exporter/collector/mongodb_collector.go:277 +0x36e
  github.com/percona/mongodb_exporter/collector.(*MongodbCollector).scrape()
      /Users/aleksi/Code/Percona/PMM-2.0/src/github.com/percona/mongodb_exporter/collector/mongodb_collector.go:213 +0x69a
  github.com/percona/mongodb_exporter/collector.(*MongodbCollector).Collect()
      /Users/aleksi/Code/Percona/PMM-2.0/src/github.com/percona/mongodb_exporter/collector/mongodb_collector.go:162 +0x46
  github.com/prometheus/client_golang/prometheus.(*Registry).Gather.func1()
      /Users/aleksi/Soft/GOPATH/pkg/mod/github.com/prometheus/client_golang@v1.0.0/prometheus/registry.go:442 +0x1f5

Goroutine 120 (running) created at:
  github.com/prometheus/client_golang/prometheus.(*Registry).Gather()
      /Users/aleksi/Soft/GOPATH/pkg/mod/github.com/prometheus/client_golang@v1.0.0/prometheus/registry.go:534 +0x11d6
  github.com/prometheus/client_golang/prometheus/promhttp.HandlerFor.func1()
      /Users/aleksi/Soft/GOPATH/pkg/mod/github.com/prometheus/client_golang@v1.0.0/prometheus/promhttp/http.go:126 +0x7d
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/Cellar/go/1.13.8/libexec/src/net/http/server.go:2007 +0x51
  github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerInFlight.func1()
      /Users/aleksi/Soft/GOPATH/pkg/mod/github.com/prometheus/client_golang@v1.0.0/prometheus/promhttp/instrument_server.go:40 +0xe4
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/Cellar/go/1.13.8/libexec/src/net/http/server.go:2007 +0x51
  github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerCounter.func1()
      /Users/aleksi/Soft/GOPATH/pkg/mod/github.com/prometheus/client_golang@v1.0.0/prometheus/promhttp/instrument_server.go:100 +0xf8
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/Cellar/go/1.13.8/libexec/src/net/http/server.go:2007 +0x51
  net/http.(*ServeMux).ServeHTTP()
      /usr/local/Cellar/go/1.13.8/libexec/src/net/http/server.go:2387 +0x288
  net/http.serverHandler.ServeHTTP()
      /usr/local/Cellar/go/1.13.8/libexec/src/net/http/server.go:2802 +0xce
  net/http.(*conn).serve()
      /usr/local/Cellar/go/1.13.8/libexec/src/net/http/server.go:1890 +0x837

Goroutine 121 (running) created at:
  github.com/prometheus/client_golang/prometheus.(*Registry).Gather()
      /Users/aleksi/Soft/GOPATH/pkg/mod/github.com/prometheus/client_golang@v1.0.0/prometheus/registry.go:534 +0x11d6
  github.com/prometheus/client_golang/prometheus/promhttp.HandlerFor.func1()
      /Users/aleksi/Soft/GOPATH/pkg/mod/github.com/prometheus/client_golang@v1.0.0/prometheus/promhttp/http.go:126 +0x7d
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/Cellar/go/1.13.8/libexec/src/net/http/server.go:2007 +0x51
  github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerInFlight.func1()
      /Users/aleksi/Soft/GOPATH/pkg/mod/github.com/prometheus/client_golang@v1.0.0/prometheus/promhttp/instrument_server.go:40 +0xe4
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/Cellar/go/1.13.8/libexec/src/net/http/server.go:2007 +0x51
  github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerCounter.func1()
      /Users/aleksi/Soft/GOPATH/pkg/mod/github.com/prometheus/client_golang@v1.0.0/prometheus/promhttp/instrument_server.go:100 +0xf8
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/Cellar/go/1.13.8/libexec/src/net/http/server.go:2007 +0x51
  net/http.(*ServeMux).ServeHTTP()
      /usr/local/Cellar/go/1.13.8/libexec/src/net/http/server.go:2387 +0x288
  net/http.serverHandler.ServeHTTP()
      /usr/local/Cellar/go/1.13.8/libexec/src/net/http/server.go:2802 +0xce
  net/http.(*conn).serve()
      /usr/local/Cellar/go/1.13.8/libexec/src/net/http/server.go:1890 +0x837
==================
```

```
go install -race ./...
mongodb_exporter --log.level=debug --collect.database --collect.collection --collect.topmetrics --collect.indexusage --collect.connpoolstats
```

```
ab -n 1000 -c 10 http://localhost:9216/metrics
```

I'm not sure if it has to be global, or if we need that functionality at all, but race condition is definitely there.

File: collector/common/collection.go (hunk 2 above)
Comment (artemgavrilov) [reply collector/common/collection.go]: Done

File: collector/common/system.go (hunk 3 above)
Comment (artemgavrilov) [reply collector/common/system.go]: Done

File: collector/mongod/collections_status.go (hunk 4 above)
Comment (artemgavrilov) [reply collector/mongod/collections_status.go]: Confirmed and fixed

File: collector/mongod/index_usage.go [hunk 5]
```diff
@@ -50,91 +50,84 @@ func (indexStats *IndexStatsList) Export(ch chan<- prometheus.Metric) {
...
-
-					delete(logSuppressIS, dbName+"."+coll.Name)
-					// Label index stats with corresponding db.collection
-					for i := 0; i < len(collIndexUsageStats.Items); i++ {
-						collIndexUsageStats.Items[i].Database = dbName
-						collIndexUsageStats.Items[i].Collection = coll.Name
-					}
-					indexUsageStatsList.Items = append(indexUsageStatsList.Items, collIndexUsageStats.Items...)
-				}
+			if err := c.Err(); err != nil {
```
Comment (percona-csalguero) [inline collector/mongod/index_usage.go]: Do we need the assignment?

Comment (artemgavrilov) [reply collector/mongod/index_usage.go]: Nope, I assume it's just code style from `mongo-go-driver` example. Should I change it? (same code all over the project)

File: collector/mongos/collections_status.go [hunk 6]
```diff
@@ -98,70 +100,58 @@ func (collStatList *CollectionStatList) Describe(ch chan<- *prometheus.Desc) {
...
-		} else {
+			continue
+		}
 
-			type collListItem struct {
-				Name string `bson:"name,omitempty"`
-				Type string `bson:"type,omitempty"`
+		logSuppressCS.Delete(dbName)
+		for _, collName := range collNames {
+			if common.IsSystemCollection(collName) {
```
Comment (BupycHuk) [inline collector/mongos/collections_status.go]: Is it possible that https://github.com/percona/mongodb_exporter/pull/177/files#diff-007e82e9748a45f49784e15a4c20541cR125 returns collection from system database?

File: collector/mongod/index_usage.go (hunk 5 above)
Comment (AlekSi) [reply collector/mongod/index_usage.go]: We need to stop somewhere with this PR :)

File: shared/sync_string_set.go [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+type SyncStringSet struct {
+	m  map[string]struct{}
+	mu sync.RWMutex
+}
+
+// NewSyncStringSet returns new SyncStringSet
+func NewSyncStringSet() *SyncStringSet {
+	return &SyncStringSet{
+		m:  make(map[string]struct{}),
+		mu: sync.RWMutex{},
```
Comment (AlekSi) [inline shared/sync_string_set.go]: Please never write `sync.RWMutex{}` or `sync.WaitGroup{}`:
1. Zero values of those types are useful. No initialisation is needed.
2. There is no struct field you can initialise.
3. In fact, it is not even important that it is a struct.

In that case, just drop that line.

File: collector/mongos/collections_status.go (hunk 6 above)
Comment (artemgavrilov) [reply collector/mongos/collections_status.go]: Yes, `ListCollectionNames` returns all collections names in current db, even system. Error raised when you try to get access to such system collection.

File: shared/sync_string_set.go (hunk 7 above)
Comment (artemgavrilov) [reply shared/sync_string_set.go]: Done

File: collector/mongod/index_usage.go (hunk 5 above)
Comment (artemgavrilov) [reply collector/mongod/index_usage.go]: Ok, next time)

File: collector/mongos/collections_status.go (hunk 6 above)
Comment (artemgavrilov) [reply collector/mongos/collections_status.go]: I mean mongo db there is system collections in non-system databases, `system.profile` for example.

Comment (artemgavrilov) [reply collector/mongos/collections_status.go]: System databases will be skipped at line https://github.com/percona/mongodb_exporter/pull/177/files#diff-007e82e9748a45f49784e15a4c20541cR121

**Pull request merged**